server/spec/hypervisor_check_in_spec.rb

```diff
@@ -295,4 +301,295 @@ def create_virtwho_client(user)
+    results1 = job_detail1['resultData']
+    job_detail2 = async_update_hypervisor(owner2, consumer2, hostname, first_guest_list)
+    results2 = job_detail2['resultData']
+    # Check in each org
+    results1.created.size.should == 1
+    results2.created.size.should == 1
+    results1.updated.size.should == 0
+    results2.updated.size.should == 0
+    results1.unchanged.size.should == 0
+    results2.unchanged.size.should == 0
+    results2.failedUpdate.size == 0
+    # Now check in each org again
+    job_detail1 = async_update_hypervisor(owner1, consumer1, hostname, first_guest_list)
+    results1 = job_detail1['resultData']
+    job_detail2 = async_update_hypervisor(owner2, consumer2, hostname, first_guest_list)
+    results2 = job_detail2['resultData']
+    # Nothing should have changed
+    results1.created.size.should == 0
+    results2.created.size.should == 0
+    results1.updated.size.should == 0
+    results2.updated.size.should == 0
+    results1.unchanged.size.should == 1
+    results2.unchanged.size.should == 1
+    results2.failedUpdate.size.should == 0
+    # Send modified data for owner 1, but it shouldn't impact owner 2 at all
+    job_detail1 = async_update_hypervisor(owner1, consumer1, hostname, second_guest_list)
+    results1 = job_detail1['resultData']
+    job_detail2 = async_update_hypervisor(owner2, consumer2, hostname, first_guest_list)
+    results2 = job_detail2['resultData']
+    # Now owner 1 should have an update, but owner two should remain the same
+    results1.created.size.should == 0
+    results2.created.size.should == 0
+    results1.updated.size.should == 1
+    results2.updated.size.should == 0
+    results1.unchanged.size.should == 0
+    results2.unchanged.size.should == 1
+    results2.failedUpdate.size.should == 0
+  end
+
+  # Tests a scenario where permissions were blocking update:
+  it 'should allow virt-who to update mappings [async]' do
+    virtwho1 = create_virtwho_client(@user)
+    hostname = random_string('my-host')
+    job_detail = async_update_hypervisor(@owner, virtwho1, hostname, ['g1', 'g2'])
+    job_detail['resultData'].created.size.should == 1
+
+    job_detail = async_update_hypervisor(@owner, virtwho1, hostname, ['g1', 'g2'])
+    job_detail['resultData'].unchanged.size.should == 1
+  end
+
+  it 'should block virt-who if owner does not match identity cert [async]' do
+    virtwho1 = create_virtwho_client(@user)
+    host = random_string('my-host')
+    guests = ['g1', 'g2']
+
+    host_mapping = get_async_host_guest_mapping(host, guests)
+    lambda do
+      job_detail = JSON.parse(virtwho1.hypervisor_update(@owner2['key'], host_mapping))
+    end.should raise_exception(RestClient::ResourceNotFound)
+  end
+
+  it 'should raise bad request exception if mapping was not provided [async]' do
+    virtwho = create_virtwho_client(@user)
+    lambda do
+      virtwho.hypervisor_update(@owner['key'], nil)
+    end.should raise_exception(RestClient::BadRequest)
+  end
```

Comment:
Really not great to see the amount of copy paste in here. At the very least could we move the async tests to directly below their non-async counterparts, so at least when something needs an edit, we can see both spots on screen at once.

I still think with a bit more work we could abstract out the API call dependent parts into methods, and sub-class the test suite for async tests that just implement those methods. (i.e. a method for setup that takes common args, and methods to verify the result json that take common args)

If you say it's not feasible a second time I'll believe you. :) It's just a LOT of copy paste.